packages/reporter/src/test/test-model.ts

```diff
@@ -89,7 +97,9 @@ export default class Test extends Runnable {
   }
 
   @computed get state () {
-    return this.lastAttempt ? this.lastAttempt.state : 'active'
+    // Use the outerStatus of the last attempt to determine overall test status, if present,
+    // as the last attempt may have 'passed', but the outerStatus may be marked as failed.
+    return this.lastAttempt ? (this.lastAttempt._testOuterStatus || this.lastAttempt.state) : 'active'
   }
```

Comment:
I think the main reason I am not a fan of doing this is it promotes somewhat of a reimplementation of the business logic in `calculateTestStatus`, which we would also need to reflect in the server reporter. I would rather `calculateTestStatus` handle the business logic for determining the outerStatus of a test which the state computation should just reference. I wonder if it makes sense to just reference `this.lastAttempt._testOuterStatus` and remove the fallback on `this.lastAttempt.state` here, which might expose if we are missing the calculation someplace.

Reply:
That makes sense. What about setting `test.state` in `calculateTestStatus` instead of `_testOuterState`? Is there a case where the outer status should/would not be the same as `test.state`?

Reply:
>  That makes sense. What about setting test.state in calculateTestStatus instead of _testOuterState? Is there a case where the outer status should/would not be the same as test.state?

There is, which is why it's probably good to think of `test.state` and `test._cypressTestStatusInfo.outerStatus` as two different properties. `test.state` within mocha refers to the `state` on the test attempt, and if the last attempt, the overall `test.state`. `mocha` makes the assumption on two things that we are changing here:

* As soon as a test attempt (retry) passes, the whole test passes (This couples the last attempt of the test to the overall status of the attempt, which is the assumption `mocha` makes.)
* passing tests cannot be re-attempted (retried)

In this case, we care about the first bulletpoint, where we need to decouple the status of the last test attempt to the status of the whole test. For example, we might have an `experimentalStrategy` of `detect-flake-but-always-fail` with `stopIfAnyPassed=true`, where the test might look like this:

```
deterministic flaky test
  ✖(Attempt 1 of 10) deterministically runs pass/fail on this test
  ✓(Attempt 2 of 10) deterministically runs pass/fail on this test
  ✖ deterministically runs pass/fail on this test
```
Where the `state` on the test is `passed` (as it was the last attempt and is coupled to the state of the attempt as well), but the `outerStatus` is `failed`. `mocha` by default wants to pass this test to make it look like this:

```
deterministic flaky test
  ✖(Attempt 1 of 10) deterministically runs pass/fail on this test
  ✓(Attempt 2 of 10) deterministically runs pass/fail on this test
  ✓deterministically runs pass/fail on this test
```

This is what the [patch](https://github.com/cypress-io/cypress/blob/feat/experimental_retries_impl/packages/driver/patches/mocha+7.0.1.dev.patch#L72) is doing. We delay the `mocha` `fail` event in this scenario to avoid mocha side effects by marking the `passed` last attempt as `failed`, but make sure `fail` the test in the [runner](https://github.com/cypress-io/cypress/blob/feat/experimental_retries_impl/packages/driver/src/cypress/runner.ts#L484). This allows us to have a last attempt that may pass, but isn't coupled to the overall test state. Hopefully I explained that well 😬

Reply:
Yeah, that makes sense. We need the extra property to make our mocha hack work 😅